enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,435 @@
+
+In addition, future phase requirements around package dependency resolution may require a new design entirely. While this is a notable short term trade-off, it allows the OLM team to continue iterating on their "deppy" component in the background. Deppy is an OLM 1.x component that's responsible for dependency resolution. A future design may elect to use deppy and it's APIs, and delegate package selection to that component. The usage of feature flags can be used to toggle the underlying implementation layer once that component is ready.
+
+This phase will likely restrict which OLM-based operators admins can install. Those requirements will be detailed below. The immediate pool of candidates may include the cert-manager and service-binding-operator operators.a
+
+#### Phase 1: Expand Capabilities
+
+Summary: Introduce depedency resolution, and expand the set of supported capabilities for admins and platform operator authors. Help establish, or improve, any first-class APIs at the installer and cluster runtime levels.
+
+The previous phase aimed to get the foundational layer in place to start onboarding more OLM-based operators to participate in the cluster's lifecycle. The notable limitations with phase 0 is requiring manual admin intervention around editing the installer's Kubernetes manifests and upgrading PO packages after installation. This phase will aim to address those limitations by introducing first-class concepts.
+
+This phase will also attempt to introduce package management and dependency resolution primitives. Candidate PO packages may place dependencies on other packages and GVKs, in addition to more granular constraints that are needed in order to accommodate EUS-EUS drift and platform-specific cluster invariants.
+
+#### Phase 2: QoL Improvements
+
+Summary: Prepare for general availability by tackling any performance improvements, shoring up the overall design/implementation/UX, and any QoL improvements that have been identified from previous phase feedback.
+
+This phase has largely been unexplored. Candidates for this phase may include:
+
+- Integration with the OCP console, hypershift, KCP, etc.
+- Improved instrumentation.
+- Improved performance of the overall stack.
+- Improved UX around the user-facing APIs.
+
+### PO Requirements
+
+- Must package Kubernetes manifests using OLM's registry+v1 bundle format.
+- Cannot specify non-AllNamespace CSV install modes.
+- Cannot declare package or GVK dependencies.
+- All package bundles must live in the redhat-operators CatalogSource.
+
+### High-Level Design
+
+In order to enable admins to express their desired intent to install an OLM-based operator in day 0 operations, a new, top-level `PlatformOperator` cluster-scoped CRD will be introduced. This resource can help serve as the bridge between the openshift installer, and the on-cluster components.
+
+The initial API design will include a spec.Package field, which references an OLM-based operators's package name:
```

Comment:
> then I would expect to be able to look at the status fields on that resource to know when the installation is complete. That's not covered here

Yep, this would be expectation as well. We've been prototyping a couple of different implementations over the last couple of sprints, and the status management typically exposes the following states:

- Was the PO manager able to successfully source/find the desired package name?
- Was the PO manager able to successfully apply the select registry+v1 bundle contents?

In either of those cases, I'd expect to see the PO manager bubble up any failure states to an individual PO resource's status so admins don't need to dive into the underlying resources that were used to facilitate/abstract these interactions away from the admin.

> Is there some reason we can't just say the PO name is the name of the thing to be installed?

I don't have a strong opinion here w.r.t the phase 0 top-level resource, and I think we have some rope here given the tech preview guidelines, so I'm fine with inferring the package name from the PO's name. In future phases, we may need to tweak this top-level resource, or even revisit it to be a simple wrapper around OLM's "Operator" API once that OLM 1.x API gets fleshed out further.

I think we still run into the same core UX issues around upfront package discovery, but those can be eased through documentation, sample PO configurations, etc, especially if the phase 0 candidate pool of supported OLM operators is small.

Thoughts @joelanford @bparees?

Reply:
using the package name as the resource name seems ok to me.  it gets us package duplication prevention for free (well, at the PO level, not at the olm level) too, so that's nice.

Reply:
Are there any restrictions on introducing a "barebones" API into the o/api repository given we can imply package naming through the PlatformOperator's custom resource metadata.Name? I don't think we have any immediate need to introduce any spec fields here, especially since the Operator API is still a WIP and we haven't really fleshed out that part of the OLM 1.x stack yet.

Reply:
> Are there any restrictions on introducing a "barebones" API into the o/api repository

i'm not aware of any explicit restrictions, but you will certainly get questions.  I would hope "this is a tech preview api" plus the general explanation would be sufficient to overcome any concerns.

Reply:
Yup, and I think that's a valid question from the API reviewer's perspective. Hopefully we have some room here given the TP guidelines. In the meantime, I can flesh out the "API Extensions" section, update the proposal's high-level design, and creating an o/api in parallel to help kick-start that conversation ahead of time.

Reply:
One use case the @elmiko brought up with CCMs is a cluster installer saying "I want a CCM operator installed. Pick the right one for me please!", thus delegating the choice to the cluster, but stating the intent to treat that choice as a PO.

If we want to support something like that, using `metadata.name` as the package name would cause problems since the cluster installer doesn't know the name in advance.

Reply:
> If we want to support something like that, using metadata.name as the package name would cause problems since the cluster installer doesn't know the name in advance.

i think a flow like that would require another controller (either in payload or a PO) that was responsible for creating the platform-specific PO install request resource.

how would you envision it working even if metadata.name didn't have to be the package name?  Something would still have to come update the spec to make it clear what PO we actually wanted.

Reply:
> One use case the @elmiko brought up with CCMs is a cluster installer saying "I want a CCM operator installed. Pick the right one for me please!", thus delegating the choice to the cluster, but stating the intent to treat that choice as a PO.

i just want to add that so far the use cases i'm envisioning would have a platform integrator specifying the platform for any installation that might happen. (eg MomAndPopCloud have created an installation supplement for OpenShift on their infra, they allow users to create clusters through a web interface or somesuch)

Reply:
Just copy-pasting my reply from slack:

> My current perspective is we can afford to make breaking changes here, and using the resource's name has some nice properties around getting package de-duplication for free (as Doug pointed out earlier in that comment chain).

Given this will be tech preview during 4.12, how much effort do we want to invest in perfecting this API now? I think we may see phase 1 requirements, and the potential use cases outlined above, influence the API design in the intermediate term anyways.

Reply:
Just wanted to add to this thread some of my concerns about using the name field to proxy the OLM package name. It has come up on another thread that the set of characters allowed in `metadata.name` is not sufficient to cover the set of characters in an OLM package name. Which means we can't create a 1:1 mapping. There was suggestion that we would have to map `_` to `-`, which, i suspect is not actually going to be intuitive and there's likely more than just that one character, need to find the requirements for an OLM package name to confirm.

If we can't provide a 1:1 mapping between `metadata.name` and the package names I think that would be a hard no from an API review perspective, creating a cipher to translate between the two is not intuitive and likely won't age well

Reply:
I just tried creating an OLM catalog that specified package names that used a "_" separator:

<details>

---
schema: olm.package
name: prometheus_operator
defaultChannel: "4.12"
---
schema: olm.channel
package: prometheus_operator
name: "4.12"
entries:
  - name: prometheus-operator.v0.1.0
---
schema: olm.bundle
name: prometheus-operator.v0.1.0
package: prometheus_operator
image: quay.io/operatorhubio/prometheus:v0.47.0
properties:
  - type: olm.package
    value:
      packageName: prometheus_operator
      version: 0.1.0

</details>

And that passed our static and runtime validation of those catalog contents. My preference for this phase would be to minimize any requisite OLM 0.x changes, and just handle this in the PlatformOperator API. It seems like there's some clear trade-offs with that approach:

- A reduced local UX experience as `k get po` may produce skew between the resource name and the configured spec.PackageName value
- We'd need to add validation to ensure that a cluster cannot contain multiple package names
- It's unclear to me whether a spec.PackageName required field will stick around for future phases, given there will be less restrictions/requirements as we expect the set of supported functionality. That seems to imply churn in the API, which seems fine to me given we're releasing under TP guidelines, and migration is out-of-scope for right now.

But using the resources name to configure the desired package name also seems fairly flawed on it's own, and there's some UX limitations in that domain as well. I guess my question is whether we're comfortable with phase 1 having a different API surface (i.e. in terms of spec fields) compared to phase 0 as we scope out requirements?

In another comment, @joelanford had ad mentioned a validation webhook may have some extra complexity that's required to validate that a cluster only has a single PO package installed. A validation webhook implementation is another candidate for something that will get replaced during phase 0 in favor of modeling cluster invariants (e.g. a single package per cluster) as dependency resolution constraints. Are we okay with churn in this domain as well?

Reply:
My other comment was:

> re: dealing with duplicate names if we use spec.packageName, it seems like we would need a mutex somewhere to guarantee avoidance of duplicates. It seems like even a validating webhook would need a mutex because you have to consider the possibility of two requests arriving simultaneously, where both have their their package names checked against the cluster prior to either being admitted.

If there's concern about what we're signing up for despite this being introduced as TP, what if we just committed right now to the entire API disappearing when we graduate to GA (e.g. if we make the group name `techpreview.platform.openshift.io`, we would be forced to throw this away and not worry about migrations and CRD conversion from TP to GA)

Reply:
That's an interesting idea. The main goal for this top-level API was to avoid requiring any direct installer/CVO component changes during this phase, while being able to proxy status back to the CVO. If we create a first-class UX next phase, then in theory, the installer component could just include the requisite rukpak/deppy/etc. manifests before cluster creation?

We'd still need a way to proxy status back to the CVO, and it sounded like in previous threads that while we want POs to act like COs, we may want to avoid having PO packages hook into the CO API for communicating their status back to the CVO. If we can solve both of those concerns, then adding this as a GA requirement seems reasonable to me.

Reply:
For the duplication, one way to avoid this would be to have a controller "accept" a PO when it's created. This controller would perform any pre-requisite checks needed (including checking for uniqueness) and then add an `Accepted` condition to the PO status.

If while reconciling the object, the operator finds a duplicated package name on an already `Accepted` PO, it would set the condition to `False` on the duplicate PO with an appropriate message.

The rest of the controllers would have a check ahead of time that skips reconciling any PO that has not been `Accepted`.

---

To make it easy to see the package name, we can add a custom printer column so that `k get po` returns a column with the package name from our metadata field.

I will note, using `po` as a shortname might cause confusion as that's already registered for pods

Reply:
Just summarizing the slack conversations + workloads arch call discussions:

- Use a dedicated spec.PackageName field for this phase given that OLM's package name schema may accept arbitrary strings, and we cannot guarantee a 1:1 mapping with a resource's name. Adding translation logic between a resource's metadata.Name and an OLM package name introduces additional complexity.
- The spec.PackageName can be a required field, and mark it optional if future phases don't need a dedicated package name anymore.
- We could bridge the gap between the local kubectl UX through printer columns that output the spec.PackageName field. That would help alleviate any skew between `kubectl get platformoperator` where the resource name doesn't match the configured package name spec field.

I can update the proposal with this implementation in the next round of edits.

---

It sounds like we got consensus on using the spec.PackageName field. Now, the missing gap becomes ensuring package name uniqueness, and it sounds like there's still some conversion remaining in that domain:

- Implementing an admission controller that checks whether a PO being created specifies a package name that already exists on the cluster. This would provide quick feedback to admins after cluster creation.
- Future phases will delegate this process to OLM's deppy component, and model the single-package-per-cluster requirement as a constraint definition. This would result in a slower feedback loop as it would require performing dependency resolution and bubbling up that status back to the PO Custom resource, but still meet the requirement.
- How much weight do we want to put into perfecting that experience given the high-level goal is being able to deploy PO packages during cluster creation? It seems like in the latter case, you're already subjecting yourself to a slower feedback loop waiting for the CVO to reflect the status of the cluster.
- We can introduce an admission controller in future phases, and ensure that the POM controller(s) handle this edge case, and bubble up the status to the resource's to ensure we're correctly proxying status back to the CVO.

Reply:
Updated the enhancement and pivoting to an explicit spec.PackageName field, and adding some rationale for why this approach was chosen over using the resource's metadata.Name. I added the admission controller implementation as an alternative, and moved validating package name uniqueness as out-of-scope for this proposal. My current thinking is we can update this proposal, and implement that validation logic if we still have runway as we head into 4.12 feature freeze, but it shouldn't block the initial design from merging.

Reply:
> I added the admission controller implementation as an alternative, and moved validating package name uniqueness as out-of-scope for this proposal.

This seems reasonable for now. I think a naive admission webhook that doesn't consider race issues would be fairly straightforward and cover 99% of what we would need. Regardless I would agree that it doesn't seem like a 4.12 TP blocker.